netlify/netlify-plugin-lighthouse
# Optionally run onSuccess

## Overview

This was a journey! 😅 

This PR adds a new environment variable which can be used to change when Lighthouse runs during a deployment. The current default is `onPostBuild`. Setting `RUN_ON_SUCCESS=true` for a site will allow Lighthouse to run after the deployment has completed, during the `onSuccess` phase. 

This will unlock:
- Faster deployments – the site no longer waits for Lighthouse to run
- More accurate testing – we'll be testing against a deploy preview, not a static server
- Unlocks testing on fancy JS sites (SPA, SSR, ISR etc)

This will block:
- Writing reports to the site files – no longer possible as the site is already deployed
- Blocking deployments for failing thresholds – the _plugin_ will report a failure, but the deployment will still count as a successful deploy
- Testing arbitrary absolute URLs

Still to do:
- Update README to reflect new output style and functionality
- Update the Netlify UI once the `onSuccess` is complete. Currently a page refresh is required to see scores.
- Better notifications around failing thresholds, with suggestions around manually rolling-back under-performing builds.
- Review scores between sites running on each event to ensure accuracy has actually improved

In future, we'll likely be making onSuccess the new default event.

## Screenshots

This is now the entire log output of a multi-report run:
<img width="882" alt="image" src="https://user-images.githubusercontent.com/720908/225633725-06845b95-d51f-418b-8891-e029f94ecbe8.png">

## Test Plan

(WIP, it would be nice to split something out into a separate doc for this…)

### onPostBuild (existing functionality)

**Using a site with a single Lighthouse report set up**
- [ ] The scores and report should complete successfully during the onPostBuild deployment phase
- [ ] Setting a lower threshold than reported should complete successfully
- [ ] Setting a higher threshold than reported should fail and block the deploy
- [ ] Setting an `output_path` should save the generated report file to the specified location
- [ ] Setting a `serveDir` should work as expected
- [ ] Should work as expected using the no-config UI-based installation
- [ ] Should allow testing of an absolute URL using `url`
- [ ] Should allow setting a custom `settings` config to tweak device preset and locale

**Using a site with multiple Lighthouse reports set up**
Same as above

### onSuccess (new functionality)

**Using a site with a single Lighthouse report set up**
- [ ] The scores and report should complete successfully during the onSuccess deployment phase
- [ ] Setting a lower threshold than reported should complete successfully
- [ ] Setting a higher threshold than reported should show as a "plugin failure"
- [ ] Setting an `output_path` should NOT attempt to save a file
- [ ] Setting a `serveDir` should NOT affect the reports
- [ ] Should work as expected using the no-config UI-based installation
- [ ] Should not allow testing of an absolute URL using `url`
- [ ] Should allow setting a custom `settings` config to tweak device preset and locale

**Using a site with multiple Lighthouse reports set up**
Same as above
